src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2236 @@
+
+constant_exprt string_constraint_generatort::constant_char(int i) const
+{
+  return from_integer(i, get_char_type());
+}
+
+unsignedbv_typet string_constraint_generatort::get_char_type() const
+{
+  if(mode==ID_C)
+    return refined_string_typet::char_type();
+  else if(mode==ID_java)
+    return refined_string_typet::java_char_type();
+  else
+    assert(false); // only C and java modes supported
+}
+
+size_t string_constraint_generatort::get_char_width() const
+{
+  if(mode==ID_C)
+    return STRING_SOLVER_C_CHAR_WIDTH;
+  else if(mode==ID_java)
+    return STRING_SOLVER_JAVA_CHAR_WIDTH;
+  else
+    assert(false); // only C and java modes supported
+}
+
+symbol_exprt string_constraint_generatort::fresh_univ_index(
+  const irep_idt &prefix)
+{
+  return string_exprt::fresh_symbol(prefix, refined_string_typet::index_type());
+}
+
+symbol_exprt string_constraint_generatort::fresh_exist_index(
+  const irep_idt &prefix)
+{
+  symbol_exprt s=string_exprt::fresh_symbol
+    (prefix, refined_string_typet::index_type());
+  index_symbols.push_back(s);
+  return s;
+}
+
+symbol_exprt string_constraint_generatort::fresh_boolean(
+  const irep_idt &prefix)
+{
+  symbol_exprt b=string_exprt::fresh_symbol(prefix, bool_typet());
+  boolean_symbols.push_back(b);
+  return b;
+}
+
+
+/*******************************************************************\
+
+Function: string_constraint_generatort::add_axioms_for_string_expr
+
+  Inputs: an expression of type string
+
+ Outputs: a string expression that is link to the argument through
+          axioms that are added to the list
+
+ Purpose: obtain a refined string expression corresponding to string
+          variable of string function call
+
+\*******************************************************************/
+string_exprt string_constraint_generatort::add_axioms_for_string_expr(
+  const exprt & unrefined_string)
+{
+  string_exprt s;
+
+  if(unrefined_string.id()==ID_function_application)
+  {
+    exprt res=add_axioms_for_function_application
+      (to_function_application_expr(unrefined_string));
+    assert(res.type()==refined_string_typet(get_char_type()));
+    s=to_string_expr(res);
+  }
+  else if(unrefined_string.id()==ID_symbol)
+    s=find_or_add_string_of_symbol(to_symbol_expr(unrefined_string));
+  else if(unrefined_string.id()==ID_address_of)
+  {
+    assert(unrefined_string.op0().id()==ID_symbol);
+    s=find_or_add_string_of_symbol(to_symbol_expr(unrefined_string.op0()));
+  }
+  else if(unrefined_string.id()==ID_if)
+    s=add_axioms_for_if(to_if_expr(unrefined_string));
+  else if(unrefined_string.id()==ID_nondet_symbol
+          || unrefined_string.id()==ID_struct)
+  {
+    // We ignore non deterministic symbols and struct
+  }
+  else if(unrefined_string.id()==ID_typecast)
+  {
+    exprt arg=to_typecast_expr(unrefined_string).op();
+    //assert(refined_string_typet::is_unrefined_string_type(
+    //  unrefined_string.type())); // its type could be simply Object
```

Comment:
I now removed it. It would not necessarily hold, for instance in a typecast of `String` to `Object`.